experimental/python/wheel.bzl

```diff
@@ -0,0 +1,159 @@
+        ctx.attr.wheel_name,
+        ctx.attr.version,
+        ctx.attr.python,
+        ctx.attr.abi,
+        ctx.attr.platform,
+    ]) + ".whl")
+
+    inputs = depset(
+        transitive = [dep[DefaultInfo].data_runfiles.files for dep in ctx.attr.deps] +
+                     [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps],
+    )
+    arguments = [
+        "--name",
+        ctx.attr.wheel_name,
+        "--version",
+        ctx.attr.version,
+        "--python",
+        ctx.attr.python,
+        "--abi",
+        ctx.attr.abi,
+        "--platform",
+        ctx.attr.platform,
+        "--out",
+        outfile.path,
+    ]
+    arguments.extend(["--restrict_package=%s" % p for p in ctx.attr.packages])
+    for input_file in inputs.to_list():
+        arguments.append("--input_file")
+        arguments.append("%s;%s" % (input_file.short_path, input_file.path))
+
+    extra_headers = []
+    if ctx.attr.author:
+        extra_headers.append("Author: %s" % ctx.attr.author)
+    if ctx.attr.author_email:
+        extra_headers.append("Author-email: %s" % ctx.attr.author_email)
+    if ctx.attr.homepage:
+        extra_headers.append("Home-page: %s" % ctx.attr.homepage)
+    if ctx.attr.license:
+        extra_headers.append("License: %s" % ctx.attr.license)
+
+    for h in extra_headers:
+        arguments.append("--header")
+        arguments.append(h)
+
+    for c in ctx.attr.classifiers:
+        arguments.append("--classifier")
+        arguments.append(c)
+
+    for r in ctx.attr.requires:
+        arguments.append("--requires")
+        arguments.append(r)
+
+    for option, requirements in ctx.attr.extra_requires.items():
+        for r in requirements:
+            arguments.append("--extra_requires")
+            arguments.append(r + ";" + option)
+
+    for name, ref in ctx.attr.console_scripts.items():
+        arguments.append("--console_script")
+        arguments.append(name + " = " + ref)
+
+    if ctx.attr.description_file:
+        description_files = ctx.attr.description_file.files.to_list()
+        arguments.append("--description_file")
+        arguments.append(description_files[0].path)
+        inputs = inputs.union(ctx.attr.description_file.files)
+
+    ctx.actions.run(
+        inputs = inputs,
+        outputs = [outfile],
+        arguments = arguments,
+        executable = ctx.executable._wheelmaker,
+        progress_message = "Building wheel",
+    )
+    return [DefaultInfo(
+        files = depset([outfile]),
+        data_runfiles = ctx.runfiles(files = [outfile]),
+    )]
+
+py_wheel = rule(
+    implementation = _py_wheel_impl,
+    attrs = {
+        "deps": attr.label_list(
+            doc = "Dependencies of this package, e.g. py_library rules",
+        ),
+        "packages": attr.string_list(
+            mandatory = True,
+            allow_empty = False,
+            doc = """\
+List of Python packages to include in the distribution.
```

Comment:
This filtering concerns me. It looks like you're taking the runfiles of a target, including all its transitive dependencies, and removing runfiles that don't match the package prefix. Isn't that giving the user an easy way to shoot themselves in the foot, by stripping out needed dependencies from the emitted result?

I'd at least be very clear in documentation about what the exact semantics are in terms of runfiles. But even then, this hazard suggests that maybe we should also provide an easy way for users to run their unit tests on an unpacked wheel in place of their normal `py_library` targets, to avoid a situation where their tests pass but their distribution is broken.

Again, not a problem that needs to be solved now, but if you agree it is in fact a problem then let's put a TODO.

Reply:
Considering that we don't want to package third-party dependencies (e.g. numpy), our options seem to be:
 1) don't follow transitive dependencies
 2) follow  dependencies but apply additional filtering

1) suffers from a problem that a simple refactoring of a rule to be multiple rules and dependecies changes what gets packaged
2) suffers from the problem you mention.

If you believe (1) is a better solution I could try to implement that instead.